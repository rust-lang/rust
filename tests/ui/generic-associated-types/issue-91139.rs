trait Foo<T> {
    type Type<'a>
    where
        T: 'a;
}

impl<T> Foo<T> for () {
    type Type<'a> = ()
    where
        T: 'a;
}

fn foo<T>() {
    let _: for<'a> fn(<() as Foo<T>>::Type<'a>, &'a T) = |_, _| ();
    //~^ ERROR `T` does not live long enough
    //~| ERROR `T` does not live long enough
    //
    // FIXME: This error is bogus, but it arises because we try to validate
    // that `<() as Foo<T>>::Type<'a>` is valid, which requires proving
    // that `T: 'a`. Since `'a` is higher-ranked, this becomes
    // `for<'a> T: 'a`, which is not true. Of course, the error is bogus
    // because there *ought* to be an implied bound stating that `'a` is
    // not any lifetime but specifically
    // "some `'a` such that `<() as Foo<T>>::Type<'a>" is valid".
}

pub fn main() {}
