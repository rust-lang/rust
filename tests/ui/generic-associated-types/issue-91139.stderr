error: `T` does not live long enough
  --> $DIR/issue-91139.rs:14:12
   |
LL |     let _: for<'a> fn(<() as Foo<T>>::Type<'a>, &'a T) = |_, _| ();
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `T` does not live long enough
  --> $DIR/issue-91139.rs:14:12
   |
LL |     let _: for<'a> fn(<() as Foo<T>>::Type<'a>, &'a T) = |_, _| ();
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

error: aborting due to 2 previous errors

